### ISSUE TITLE
Pre-Built Binaries for Apple Silicon?

### ISSUE THREAD

**Issue by bdkjones** [sass/dart-sass#1125]
Apologies if this has been asked; I *did* search!

With the first Apple Silicon Macs about 30-45 days out, is there a plan to make the Mac pre-built downloads Universal binaries? I understand that I can try building Dart Sass on my own to support Apple Silicon, but I'd rather leave it to the experts if you're going to continue releasing pre-built binaries that support the Mac platform. Thanks!

**Comment by nex3** [comment]
In order for this to be possible, we'll need to have support for targeting this platform with the Dart SDK which I believe is tracked by https://github.com/dart-lang/sdk/issues/42774.

**Comment by bdkjones** [comment]
Makes sense. I don't see much discussion about progress on dart-lang/sdk#42774. Hopefully the slowdown running X64 through Rosetta won't be too noticeable until Dart can get this support baked in.

**Comment by nex3** [comment]
Feel free to :+1: that issue to maybe help draw attention to it!

**Comment by bdkjones** [comment]
Since I run Dart Sass in a separate process, the lack of a native version isn't a blocker for me to recompile my app as a Universal binary. That said, I'm sure there will be some performance penalty running x86 code through Rosetta. (Although, hell, with the insane benchmark numbers the M1 chip is putting up, it might be faster than native x86!)

**Comment by bdkjones** [comment]
I picked up the M1 MacBook Pro. The bottom line is that Dart Sass running through Rosetta2 is "fast enough". 

I tested it with a stylesheet that contains a few thousand lines across lots of imported files and compiling took about 0.5 seconds. There's virtually no perceptible difference.

That's not to say native support shouldn't be done—it'll only get faster.

**Comment by o-alquimista** [comment]
This also applies to Raspberry Pi 4, right? Both seem to be ARM64.

**Comment by bdkjones** [comment]
@nex3 Any inside info to share on Dart’s progress with Apple Silicon support? 

dart-lang/sdk/issues/42774 appears dead.

While Dart-Sass runs quick enough with Rosetta, there *is* a noticeable delay the first time you run it. Rosetta translates the executable on first-run, then caches it.

It would be great to know someone on the Dart team has a plan and an ETA.

**Comment by nex3** [comment]
Unfortunately, I have no particular extra insight to the Dart team's working—even when I was on the team, the VM portion of it was kind of a black box.

**Comment by bdkjones** [comment]
@nex3 gotcha. Any idea who might be a good person to ping for more info?

**Comment by nex3** [comment]
@mraleph might be able to help?

**Comment by mraleph** [comment]
I answered on the bug. We are currently blocked there but hope to be unblocked.

**Comment by bdkjones** [comment]
@mraleph @nex3 It looks like Rosetta might be going away in some parts of the world with the next update to macOS: https://twitter.com/SteveMoser/status/1366904041243041794?s=20

Those folks can fall back on the JavaScript implementation of Sass, I guess, but it would sure be nice to have Dart on board.

**Comment by a4z** [comment]
There seems to be activity on the dart M1 topic 
https://github.com/dart-lang/sdk/issues/42773

**Comment by bdkjones** [comment]
@nex3 Seems like Dart now supports ARM64, but the binary shipped with Dart Sass 1.39.2 is still x86_64 only. Any idea when those will go universal? More ARM Macs are coming in a month!

**Comment by nex3** [comment]
> Can GitHub Actions' `macos-11` run the ARM64 Dart SDK? 

I've run a test and it looks like it unfortunately does not. Until GitHub offers an ARM64 builder or Dart adds support for cross-compilation (https://github.com/dart-lang/sdk/issues/28617), the best we can do is offer script snapshots for ARM64. I'm not sure this would actually be faster than running an AOT snapshot that need to be transpiled. Is it even worth offering it as a download option at that point?

**Comment by bdkjones** [comment]
No, I don’t think so. It’s the Dart executable that needs to be Universal, not the Sass snapshot.

**Comment by nex3** [comment]
To be clear, we would be bundling the ARM64 Dart executable *with* the script snapshot, so Dart at least would be running natively. The question is whether the speed of ARM64 Dart balances out the slowness of a script snapshot.

**Comment by bdkjones** [comment]
Oh, that would be very useful. I could lipo the ARM and Intel executables into a Universal binary manually then.

I would imagine the native ARM executable will eliminate the 1-2 second hang the first time Dart Sass runs on an M1.

-Bryan

**Comment by nex3** [comment]
The package we release for Intel and the package we could release for ARM wouldn't really have the same structure, even though the files in them would have the same names. Because we're able to natively compile Sass on Intel, the `sass.snapshot` file contains a bunch of x86 machine code (an "AOT snapshot") and the `dart` file contains a minimal amount of bootstrapping necessary to load and run that machine code. On ARM, though, `sass.snapshot` would only contain a compressed representation of the Dart code, and the `dart` file would contain the full Dart virtual machine. There's not really a way to combine them to create a single universal binary.

If you're willing to do some build steps yourself, you basically have three options:

1. Compile a script snapshot by checking out Dart Sass, running `dart pub get`, and then running `dart pub run grinder pkg-compile-snapshot`. Create a universal Dart binary by downloading the x64 and arm64 Dart SDKs, and use it to invoke that snapshot. This is pretty easy to set up, but the downside is it'll be slower than it needs to be, particularly on non-ARM machines where it'll be decidedly slower than the package users could download from GitHub.

2. Compile AOT snapshots for both x64 and arm64. To do this you'll need to run `dart pub run grinder pkg-compile-native` twice, once with an x64 Dart SDK and once with an arm64 SDK. Then create a simple shell script wrapper to choose which of these to invoke based on the current OS architecture. This is obviously more complicated, but it'll be as fast as the GitHub release.

3. Compile standalone native executables for both x64 and arm64 and `lipo` them together to get a standalone universal binary. You'll need to manually invoke `dart compile exe bin/sass.dart` with both the x64 and arm64 Dart SDKs and then run `lipo` to combine them, but this will produce the fastest *and* standalonest option available. However, it has a huge downside: if you try to distribute the executable, you'll need to [code sign](https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Introduction/Introduction.html) it or users will get a big scary message when they run it. That's why we don't distribute standalone executables on Windows or Mac OS.

It's also worth noting that if users install Dart Sass via Homebrew, they'll automatically get a natively-compiled version and it'll use ARM64 if available. So that's definitely the best option if it's feasible, although IIRC you want to bundle Sass with an editor which makes that trickier.

**Comment by bdkjones** [comment]
Thanks for the great explanation! Number 3 is definitely the way to go in my case. All executable code in my app's bundle is signed as part of the distribution process, so that's not a roadblock.

Is there any configuration I need to worry about if building on my own? A 14,341 line make file, perhaps?

**Comment by nex3** [comment]
We try to keep the build process short and sweet. Just `dart pub get` and `dart compile exe bin/sass.dart`. If you want an accurate `--version` flag, you'll also need to pass `-Dversion=1.40.1` or whatever, but it'll still build without that.

**Comment by hez** [comment]
👍

**Comment by nex3** [comment]
This is now done!

**Comment by bdkjones** [comment]
Question: is the `sass.snapshot` file platform-dependent? I compared the arm and x86 versions and they had some differences.

I know the `dart` executable is platform-specific, but wondering if I can lipo those together and run a single `sass-snapshot`.

Thanks!

**Comment by nex3** [comment]
Yes, `sass.snapshot` should be assumed to be platform-specific. (It may not _always_ be, but even in those cases we reserve the right to make it platform-dependent in the future.)
